scene/gui/button.cpp

```diff
@@ -213,146 +262,160 @@ void Button::_notification(int p_what) {
+						} break;
+					}
 				}
-				icon_size = _fit_icon_size(icon_size);
 
-				if (vertical_icon_alignment == VERTICAL_ALIGNMENT_TOP) {
-					voffset = -(valign - icon_size.y) / 2;
-				}
-				if (vertical_icon_alignment == VERTICAL_ALIGNMENT_BOTTOM) {
-					voffset = (valign - icon_size.y) / 2 + text_buf->get_size().y;
-				}
+				Rect2 icon_region = Rect2(icon_ofs, icon_size);
 
-				if (icon_align_rtl_checked == HORIZONTAL_ALIGNMENT_LEFT) {
-					icon_region = Rect2(style_offset + Point2(icon_ofs_region, voffset + Math::floor((valign - icon_size.y) * 0.5)), icon_size);
-				} else if (icon_align_rtl_checked == HORIZONTAL_ALIGNMENT_CENTER) {
-					icon_region = Rect2(style_offset + Point2(icon_ofs_region + Math::floor((size.x - icon_size.x) * 0.5), voffset + Math::floor((valign - icon_size.y) * 0.5)), icon_size);
-				} else {
-					icon_region = Rect2(style_offset + Point2(icon_ofs_region + size.x - icon_size.x, voffset + Math::floor((valign - icon_size.y) * 0.5)), icon_size);
+				if (icon_region.size.width > 0.0f) {
+					draw_texture_rect(_icon, icon_region, false, icon_modulate_color);
 				}
 
-				if (icon_region.size.width > 0) {
-					Rect2 icon_region_rounded = Rect2(icon_region.position.round(), icon_region.size.round());
-					draw_texture_rect(_icon, icon_region_rounded, false, color_icon);
+				{ // Updates the remaining drawable size.
+					if (icon_align_rtl_checked != HORIZONTAL_ALIGNMENT_CENTER) {
+						drawable_size_remained.width -= icon_size.width;
+						if (!xl_text.is_empty()) {
+							drawable_size_remained.width -= h_separation;
+						}
+					}
+					if (vertical_icon_alignment != VERTICAL_ALIGNMENT_CENTER) {
+						drawable_size_remained.height -= icon_size.height;
+					}
 				}
 			}
 
-			Point2 icon_ofs = !_icon.is_null() ? Point2(icon_region.size.width + theme_cache.h_separation, 0) : Point2();
-			if (align_rtl_checked == HORIZONTAL_ALIGNMENT_CENTER && icon_align_rtl_checked == HORIZONTAL_ALIGNMENT_CENTER) {
-				icon_ofs.x = 0.0;
-			}
+			// Draw the text.
+			if (!xl_text.is_empty()) {
+				text_buf->set_alignment(align_rtl_checked);
 
-			int text_clip = size.width - style->get_minimum_size().width - icon_ofs.width;
-			if (_internal_margin[SIDE_LEFT] > 0) {
-				text_clip -= _internal_margin[SIDE_LEFT] + theme_cache.h_separation;
-			}
-			if (_internal_margin[SIDE_RIGHT] > 0) {
-				text_clip -= _internal_margin[SIDE_RIGHT] + theme_cache.h_separation;
-			}
-
-			text_buf->set_width(is_clipped ? text_clip : -1);
+				text_buf->set_width(-1.0f); // Calculates the width of the text without being constrained.
+				if (is_clipped) {
+					text_buf->set_width(MAX(MIN(drawable_size_remained.width, text_buf->get_size().width), 1.0f)); // Computes the constrained text width.
+				}
 
-			int text_width = MAX(1, is_clipped ? MIN(text_clip, text_buf->get_size().x) : text_buf->get_size().x);
+				const Size2 text_buf_size = text_buf->get_size();
```

Comment:
@YuriSizov There seems to be some change in the behavior of `text_buf->set_width()`. It can no longer affect the result of `text_buf->get_size()`.